org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/SourceFolderUpdater.java

```diff
@@ -68,19 +69,40 @@ private void updateClasspath(IProgressMonitor monitor) throws JavaModelException
         updateClasspath(newClasspathEntries, monitor);
     }
 
-    private List<IClasspathEntry> collectGradleSourceFolders() {
+    private List<IClasspathEntry> collectGradleSourceFolders() throws JavaModelException {
+        final List<IClasspathEntry> rawClasspath = ImmutableList.copyOf(this.project.getRawClasspath());
         List<IClasspathEntry> sourceFolders = FluentIterable.from(this.sourceFolders).transform(new Function<OmniEclipseSourceDirectory, IClasspathEntry>() {
 
             @Override
             public IClasspathEntry apply(OmniEclipseSourceDirectory directory) {
                 IFolder sourceDirectory = SourceFolderUpdater.this.project.getProject().getFolder(Path.fromOSString(directory.getPath()));
                 FileUtils.ensureFolderHierarchyExists(sourceDirectory);
-                IPackageFragmentRoot root = SourceFolderUpdater.this.project.getPackageFragmentRoot(sourceDirectory);
+                final IPackageFragmentRoot root = SourceFolderUpdater.this.project.getPackageFragmentRoot(sourceDirectory);
                 IClasspathAttribute fromGradleModel = JavaCore.newClasspathAttribute(CLASSPATH_ATTRIBUTE_FROM_GRADLE_MODEL, "true");
+                
+                IPath[] includes = new IPath[] {};
+                IPath[] excludes = new IPath[] {};
+
+                // preserve the includes/excludes defined by the user. 
+                // this should eventually be provided by the Tooling API,
+                // see Bug 470071
+                Optional<IClasspathEntry> userDefinedClasspathEntry = FluentIterable.from(rawClasspath).firstMatch(new Predicate<IClasspathEntry>() {
+                    
+                    @Override
+                    public boolean apply(IClasspathEntry entry) {
+                        return root.getPath().toString().equals(entry.getPath().toString());
```

Comment:
No need to compare the strings; can be replaced with:  
return root.getPath().equals(entry.getPath());